skimage/segmentation/tests/test_join.py

```diff
@@ -91,3 +92,57 @@ def test_relabel_sequential_dtype():
     assert_array_equal(fw, fw_ref)
     inv_ref = np.array([0, 0, 0, 0, 0, 1,  5,  8, 42, 99])
     assert_array_equal(inv, inv_ref)
+
+
+@pytest.mark.parametrize('dtype', (np.byte, np.short, np.intc, np.int_,
+                                   np.longlong, np.ubyte, np.ushort,
+                                   np.uintc, np.uint, np.ulonglong))
+@pytest.mark.parametrize('data_already_sequential', (False, True))
+def test_relabel_sequential_int_dtype_stability(data_already_sequential,
+                                                dtype):
+    if data_already_sequential:
+        ar = np.array([1, 3, 0, 2, 5, 4], dtype=dtype)
+    else:
+        ar = np.array([1, 1, 5, 5, 8, 99, 42, 0], dtype=dtype)
+    assert all(a.dtype == dtype for a in relabel_sequential(ar))
+
+
+def test_relabel_sequential_int_dtype_overflow():
+    ar = np.array([1, 3, 0, 2, 5, 4], dtype=np.uint8)
+    offset = 254
+    ar_relab, fw, inv = relabel_sequential(ar, offset=offset)
+    assert all(a.dtype == np.uint16 for a in (ar_relab, fw, inv))
+    ar_relab_ref = ar.astype(np.int, copy=True)
+    ar_relab_ref[ar_relab_ref > 0] += offset - 1
+    assert_array_equal(ar_relab, ar_relab_ref)
+
+
+def test_relabel_sequential_negative_values():
+    ar = np.array([1, 1, 5, -5, 8, 99, 42, 0])
+    with pytest.raises(ValueError):
+        relabel_sequential(ar)
```

Comment:
Yikes, I did not realise just how bad the results are right now when you have negative labels:

```python
In [3]: segmentation.relabel_sequential(np.array([-1, 0, 2, 2, 3, 8, -1]))
Out[3]: 
(array([4, 0, 2, 2, 3, 4, 4]),
 array([0, 0, 2, 3, 0, 0, 0, 0, 4]),
 array([-1,  0,  2,  3,  8]))
```

Note that 4 ends up representing both 8 and -1!

Reply:
Personally, though, I might offset the labels with the minimum value + 1 if there are negative labels, and then proceed as normal. Thoughts?

Reply:
I don't think that works with the current semantics of the function that the input can contain any negative values, since `label_field` can be used to index into `fw`. From the example in the docstring:
```python
>>> relab, fw, inv = relabel_sequential(label_field)
>>> (fw[label_field] == relab).all()
```

Reply:
A very good point! Ignore me! =P